### PR TITLE
Fix #1547, parts of #169, #5344, #5365, #5411: Add data layer support for (multiple) classrooms & topic dependencies, and prepare for #4885

### LINKED ISSUE #1547: GaeSolution's 'correct_answer' field should be 'Any?' instead of 'String?'
**Is your feature request related to a problem? Please describe.**
According to [state_domain.py](https://github.com/oppia/oppia/blob/develop/core/domain/state_domain.py#L324), `correct_answer` in GaeSolution.kt should not be typecast as 'String' but as 'Any'.  

**Describe the solution you'd like**
The type declared in `GaeSolution.kt` for `correct_answer` must be changed to `Any?`. On top of this, in `exploration.proto` we must cast `correct_answer` to be an InteractionObject and must change `StateRetreiver.kt` in order to parse the field correctly.

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation
Fixes #1547
Fixes part of #169
Fixes part of #5344
Fixes part of #5365
Fixes part of #5411

The main purpose of this PR is to introduce support for multiple classrooms in the data layer of the app (with minimal domain integration to avoid the change extending beyond the lesson structures). However, the PR is also doing a few more things including preparing the Android codebase for introducing an asset download script (#4885) and other peripheral cleanups of code (rather than updating it) that was noticed along the way.

### Preparing for multiple classrooms

This addresses part of #5365.

#5344 is introducing support for multiple classrooms in the app. To help prepare for those changes, this PR introduces the necessary data structure and domain loading changes to load a new proto structure:

```proto
message ClassroomList {
  repeated ClassroomRecord classrooms = 1;
}
message ClassroomRecord {
  string id = 1;
  map<string, TranslationMapping> written_translations = 2;
  SubtitledHtml translatable_title = 3;
  LessonThumbnail classroom_thumbnail = 4;
  map<string, TopicIdList> topic_prerequisites = 5;
  message TopicIdList {
    repeated string topic_ids = 1;
  }
}
```

Rather than just a flat topic list. Some important details to note:
- The recommended topics structure has been updated to use this new ``topic_prerequisites`` value being loaded from classrooms. This will also extend to production assets since the asset download script from #4885 is also being updated to include support for this multiple classrooms structure to address the remainder of #5365.
- To minimize domain changes, the new loading code assumes only **one** classroom is present. TODOs have been added on #5344 to extend this to support multiple classrooms.
- Current loading code (including for JSON) is ignoring all but: ``classrooms``, ``id``, and ``topic_prerequisites`` (including ``topic_ids``) from the protos above. These other fields are expected to have supported added as part of #5344.
- There were some color simplifications made in ``TopicListController``. These largely shouldn't have a major impact outside of developer assets. These changes were made to ensure non-specificity to the previous lesson classroom. In general, all of this should eventually be removed in favor of loading colors from lesson assets, but that will need to wait until the JSON pipeline is completely removed.

### Asset priming removal

This addresses part of #169.

``PrimeTopicAssetsController`` and its implementation were responsible for hackily pre-loading all lesson images and audio to be on-device to enable offline support. This was the first attempt at offline support early in the app's development, but it had a few significant drawbacks:
- It required preloading everything upon first app open. Since it can take a while for loading to occur, some robustness needed to be built in for pausing, cancelling, and resuming. I'm not certain if these were even 100% handled in the current implementation.
- It doesn't perform strong compatibility checks until you're in the app which means lesson incompatibilities would just cause the app to get stuck rather than addressing it during lesson import time (e.g. via an asset downloader script).
- It required very significant workarounds to existing loading pipelines that aren't ideal to keep in the codebase long-term (code smell).
- There's no guarantee the user even has enough disk space to download all the needed assets (particularly audio), or if they'll have sufficient internet connectivity & bandwidth to perform those downloads upon first app open.

This approach was abandoned after the earliest alpha releases for an asset download script (which is now being migrated over to this codebase per #5411.

This removal unfortunately required removing a module that was referenced in a lot of tests throughout the codebase. While the removal itself was fairly simple, it does affect a lot of files.

Other areas changed (but unaffected by tests since these flows didn't have automated tests):
- ``SplashActivityPresenter`` for enabling the downloader to start and block the UI using a dialog box while the downloading occurred.
- ``AudioPlayerController` for removing the special loading logic for primed audio files (the app now no longer supports loading audio files from disk as we don't yet have a good long-term solution for offline-available audio files due to their significant size).
- ``GlideImageLoader`` for removing support for loading locally cached images (only through this flow; the flow we use for the asset download script uses a different annotation and is still supported).

As alluded to above, two annotations were removed as part of this cleanup:
- ``CacheAssetsLocally``: a boolean indicating whether the prime download flow should be enabled.
- ``TopicListToCache``: this specified the list of topics to pre-download if the flow was enabled.

### GAE structure cleanup & preparing for asset script

Preparing for #4885 included a few other changes, some of which were nice-to-have:
- Introducing support for incorporating the protos from https://github.com/oppia/oppia-proto-api (specifically https://github.com/oppia/oppia-proto-api/pull/1 since they are still a work-in-progress).
- Adding ``java_proto`` versions for many of the app's proto structures (since the download script runs in the JVM and doesn't use the javalite proto environment).
- Removed all of the unused GAE services and models (addressing #1547 by essentially making it obsolete), plus their mocks. These were never hooked up, and they're never going to be: the long-term plan for the codebase is to use new proto endpoints that will be added to Oppia web. These Retrofit endpoints have actually been rebuilt and repurposed to be used in the asset download script as a medium-term stop-gap until the permanent proto endpoints can be added to Oppia web.
- ``RemoteAuthNetworkInterceptorTest`` was updated to use a different example service since ``TopicService`` has been removed. The services for platform parameters and user feedback are being kept.
- Some test file & KDoc exemptions have been removed since their corresponding files have been deleted.

Note that the new Java proto targets & oppia-proto-api targets aren't being used yet, but they will be in future PRs. This just provides the basis of support for the asset download script while also helping to split up the code to review across multiple PRs.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
This is essentially only data infrastructural except for a couple of points:
- Topic loading is now happening through a classrooms structure rather than a tropic ID list. Since there's only one test & one production classroom, this shouldn't actually change the UX.
- An at-app-open flow for predownloading image & audio assets has been removed. This hasn't been used since the very earliest alpha releases of the app, so it won't actually affect any users.
- Some colors for developer lesson and topic thumbnails were updated--see above.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -15,6 +15,22 @@ android_sdk_repository(
     build_tools_version = "29.0.2",
 )
 
+# Oppia's backend proto API definitions.
```
Comment (BenHenning) [inline WORKSPACE:18]: Document this.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt [hunk 2]
```diff
@@ -56,7 +55,6 @@ class SplashActivityPresenter @Inject constructor(
   private val activity: AppCompatActivity,
   private val oppiaLogger: OppiaLogger,
   private val appStartupStateController: AppStartupStateController,
-  private val primeTopicAssetsController: PrimeTopicAssetsController,
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt:59]: Document this.

File: domain/src/main/java/org/oppia/android/domain/topic/PrimeTopicAssetsController.kt [hunk 3]
```diff
@@ -1,18 +0,0 @@
-package org.oppia.android.domain.topic
-
-// TODO(#169): Remove this controller & download priming once downloads are properly supported.
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/PrimeTopicAssetsController.kt:3]: Update docs.

Also, explain removal of this controller & its impl, plus the annotations.

File: data/src/main/java/org/oppia/android/data/backends/gae/model/BUILD.bazel [hunk 4]
```diff
@@ -9,44 +9,12 @@ load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kt_android_library")
 kt_android_library(
     name = "model",
     srcs = [
-        "GaeAnswerGroup.kt",
```
Comment (BenHenning) [inline data/src/main/java/org/oppia/android/data/backends/gae/model/BUILD.bazel:11]: Document these removals.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt [hunk 5]
```diff
@@ -57,7 +57,7 @@ class RemoteAuthNetworkInterceptorTest {
 
   private lateinit var client: OkHttpClient
 
-  private lateinit var topicService: TopicService
+  private lateinit var platformParameterService: PlatformParameterService
```
Comment (BenHenning) [inline data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:60]: Document this change.

File: domain/src/main/assets/classrooms.textproto [hunk 6]
```diff
@@ -0,0 +1,31 @@
+classrooms {
```
Comment (BenHenning) [inline domain/src/main/assets/classrooms.textproto:1]: Document this & the JSON file.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt [hunk 7]
```diff
@@ -156,37 +150,7 @@ class AudioPlayerController @Inject constructor(
 
   private fun prepareDataSource(url: String) {
     try {
-      if (cacheAssetsLocally && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:159]: Document this removal specifically.

File: model/src/main/proto/BUILD.bazel [hunk 8]
```diff
@@ -101,24 +101,30 @@ java_lite_proto_library(
     deps = [":interaction_object_proto"],
 )
 
+java_proto_library(
```
Comment (BenHenning) [inline model/src/main/proto/BUILD.bazel:104]: Document new java_proto library changes.

File: model/src/main/proto/topic.proto [hunk 9]
```diff
@@ -547,10 +547,38 @@ message EphemeralRevisionCard {
...
+  // before they play the topic given by the key. Note that the keys of this map indicate the
+  // complete list of topics contained within this classroom. The prerequisite list of topics may
+  // include topics outside of this classroom.
+  map<string, TopicIdList> topic_prerequisites = 5;
+
+  // Represents a list of topic IDs (to be used in the context of topic deps in a classroom).
+  message TopicIdList {
+    // A list of topics IDs.
+    repeated string topic_ids = 1;
+  }
```
Comment (BenHenning) [inline model/src/main/proto/topic.proto:581]: Document new proto structures.

File: scripts/assets/kdoc_validity_exemptions.textproto [hunk 10]
```diff
@@ -283,14 +283,7 @@ exempted_file_path: "app/src/sharedTest/java/org/oppia/android/app/utility/Drawa
 exempted_file_path: "app/src/sharedTest/java/org/oppia/android/app/utility/EspressoTestsMatchers.kt"
 exempted_file_path: "app/src/sharedTest/java/org/oppia/android/app/utility/OrientationChangeAction.kt"
 exempted_file_path: "app/src/sharedTest/java/org/oppia/android/app/utility/TabMatcher.kt"
-exempted_file_path: "data/src/main/java/org/oppia/android/data/backends/gae/api/ClassroomService.kt"
```
Comment (BenHenning) [inline scripts/assets/kdoc_validity_exemptions.textproto:286]: Mention removals for here & test_file_exemptions.

File: utility/src/main/java/org/oppia/android/util/parser/image/GlideImageLoader.kt [hunk 11]
```diff
@@ -101,11 +99,6 @@ class GlideImageLoader @Inject constructor(
   }
 
   private fun loadImage(imageUrl: String): Any = when {
-    cacheAssetsLocally -> object : ImageAssetFetcher {
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/android/util/parser/image/GlideImageLoader.kt:104]: Document this removal specifically.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 12]
```diff
@@ -154,13 +150,11 @@ class TopicListController @Inject constructor(
   }
 
   private fun loadTopicListFromJson(contentLocale: OppiaLocale.ContentLocale): TopicList {
-    val topicIdJsonArray = jsonAssetRetriever
-      .loadJsonFromAsset("topics.json")!!
-      .getJSONArray("topic_id_list")
+    val topicIdList = loadCombinedClassroomTopicList()
     val topicListBuilder = TopicList.newBuilder()
-    for (i in 0 until topicIdJsonArray.length()) {
+    for (i in topicIdList.indices) {
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: Iterate the topic IDs directly, instead.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 13]
```diff
@@ -527,31 +519,12 @@ class TopicListController @Inject constructor(
...
-      RATIOS_TOPIC_ID -> listOf()
-      // Addition and Subtraction (depends on Place Values)
-      // Multiplication (depends on Addition and Subtraction)
-      // Division (depends on Multiplication)
-      // Expressions and Equations (depends on A+S, Multiplication, Division)
-      // Decimals (depends on A+S, Multiplication, Division)
-      else -> listOf()
-    }
-  }
+  private fun retrieveTopicDependencies(topicId: String): List<String> =
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:525]: Explain this change.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 14]
```diff
@@ -171,12 +165,10 @@ class TopicListController @Inject constructor(
...
 
   private fun computeComingSoonTopicList(): ComingSoonTopicList {
-    val topicIdJsonArray = jsonAssetRetriever
-      .loadJsonFromAsset("topics.json")!!
-      .getJSONArray("topic_id_list")
+    val topicIdList = loadCombinedClassroomTopicList()
     val comingSoonTopicListBuilder = ComingSoonTopicList.newBuilder()
-    for (i in 0 until topicIdJsonArray.length()) {
-      val upcomingTopicSummary = createUpcomingTopicSummary(topicIdJsonArray.optString(i)!!)
+    for (i in topicIdList.indices) {
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: Iterate the topic IDs directly, instead.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 15]
```diff
@@ -830,6 +793,53 @@ class TopicListController @Inject constructor(
       .setChapterPlayState(nextChapterProgress?.chapterPlayState ?: ChapterPlayState.NOT_STARTED)
       .build()
   }
+
+  // TODO(#5344): Remove this in favor of per-classroom data handling.
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:796]: Mention the temporary setup here & the limitations in expecting only a single classroom currently.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 16]
```diff
@@ -830,6 +793,53 @@ class TopicListController @Inject constructor(
...
+    return if (loadLessonProtosFromAssets) {
+      return assetRepository.loadProtoFromLocalAssets(
+        assetName = "classrooms",
+        baseMessage = ClassroomList.getDefaultInstance()
+      ).classroomsList.single() // Only one record is currently expected.
+    } else loadClassroomFromJson()
+  }
+
+  // TODO(#5344): Remove this in favor of per-classroom data handling.
+  private fun loadClassroomFromJson(): ClassroomRecord {
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:807]: Explain the limitations in the JSON format currently supported.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 17]
```diff
@@ -968,21 +978,21 @@ internal fun createChapterThumbnail3(): LessonThumbnail {
...
-    .setBackgroundColorRgb(Color.parseColor(DUCK_BG_COLOR))
+    .setBackgroundColorRgb(Color.parseColor(CHAPTER_BG_COLOR_2))
     .build()
 }
 
 internal fun createChapterThumbnail8(): LessonThumbnail {
   return LessonThumbnail.newBuilder()
     .setThumbnailGraphic(LessonThumbnailGraphic.DUCK_AND_CHICKEN)
-    .setBackgroundColorRgb(Color.parseColor(DUCK_BG_COLOR))
+    .setBackgroundColorRgb(Color.parseColor(CHAPTER_BG_COLOR_1))
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:994]: Revert these color changes.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 18]
```diff
@@ -44,8 +46,6 @@ import javax.inject.Singleton
 private const val ONE_WEEK_IN_DAYS = 7
 
 private const val TOPIC_BG_COLOR = "#C6DCDA"
-private const val BAKER_BG_COLOR = "#0F63A3"
-private const val DUCK_BG_COLOR = "#05538F"
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:48]: Revert these color changes.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed all changes.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed 2 changed files after merge (due to conflicts from develop changes).

File: domain/src/main/assets/classrooms.textproto (hunk 6 above)
Comment (BenHenning) [reply domain/src/main/assets/classrooms.textproto:1]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 12 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 13 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:525]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 14 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 15 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:796]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 16 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:807]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 17 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:994]: Kept these & explained them, instead.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 18 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:48]: Kept these & explained them, instead.

File: model/src/main/proto/topic.proto (hunk 9 above)
Comment (BenHenning) [reply model/src/main/proto/topic.proto:581]: Done.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt (hunk 2 above)
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt:59]: Done.

File: domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt (hunk 7 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/audio/AudioPlayerController.kt:159]: Done.

File: utility/src/main/java/org/oppia/android/util/parser/image/GlideImageLoader.kt (hunk 11 above)
Comment (BenHenning) [reply utility/src/main/java/org/oppia/android/util/parser/image/GlideImageLoader.kt:104]: Done.

File: WORKSPACE (hunk 1 above)
Comment (BenHenning) [reply WORKSPACE:18]: Done.

File: domain/src/main/java/org/oppia/android/domain/topic/PrimeTopicAssetsController.kt (hunk 3 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/topic/PrimeTopicAssetsController.kt:3]: Done.

File: data/src/main/java/org/oppia/android/data/backends/gae/model/BUILD.bazel (hunk 4 above)
Comment (BenHenning) [reply data/src/main/java/org/oppia/android/data/backends/gae/model/BUILD.bazel:11]: Done.

File: data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt (hunk 5 above)
Comment (BenHenning) [reply data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:60]: Done.

File: model/src/main/proto/BUILD.bazel (hunk 8 above)
Comment (BenHenning) [reply model/src/main/proto/BUILD.bazel:104]: Done.

File: scripts/assets/kdoc_validity_exemptions.textproto (hunk 10 above)
Comment (BenHenning) [reply scripts/assets/kdoc_validity_exemptions.textproto:286]: Done.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes (bumping oppia-proto-api version so that CI doesn't try to pull a non-existent version of zlib).

**Comment by BenHenning** [comment]
Ah excellent. CI is now passing! @adhiamboperes PTAL.

**Review by adhiamboperes — APPROVED** [review]
Thanks @BenHenning! The cleanup looks spiff and I have no concerns with the new additions.

**Comment by BenHenning** [comment]
Thanks @adhiamboperes! I really appreciate the review. :)

Updating to latest develop & enabling auto-merge.

**Comment by BenHenning** [comment]
Also @theMr17 I forgot to ask you to review this, so I suggest taking a look at it post-merge and let me know if there are any additions needed for multiple classrooms that won't already be included in your project.

**Pull request merged**